src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp

```diff
@@ -0,0 +1,146 @@
+#include "qgsmeshdataprovidertemporalcapabilities.h"
+
+
+QgsMeshDataProviderTemporalCapabilities::QgsMeshDataProviderTemporalCapabilities(): QgsDataProviderTemporalCapabilities()
+{}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndexTimeInMilliseconds( int group, qint64 timeSinceGlobalReference ) const
+{
+  const QDateTime &groupReference = mGroupsReferenceDateTime[group];
+
+  const qint64 timeSinceGroupeReference =
+    timeSinceGlobalReference - mGlobalReferenceDateTime.msecsTo( groupReference );
+  const QList<qint64> &datasetTimes = mDatasetTimeSinceGroupReference[group];
+  // No dataset or time too little, return invalid index
+  if ( datasetTimes.isEmpty() )
+    return QgsMeshDatasetIndex();
+
+  // If time is less than the time of the first dataset, return the first dataset
+  // Allow to display "static" dataset (as terrain elevation) whatever the time
+  if ( timeSinceGroupeReference < datasetTimes.first() )
+    return QgsMeshDatasetIndex( group, 0 );
+  for ( int i = 1; i < datasetTimes.count(); ++i )
+  {
+    if ( timeSinceGroupeReference < datasetTimes.at( i ) )
+      return QgsMeshDatasetIndex( group, i - 1 );
+  }
+
+  // no time found because time too big --> return the last index;
+  return QgsMeshDatasetIndex( group, datasetTimes.count() - 1 );
+}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndex( int group, double timeSinceGlobalReferenceInHours ) const
+{
+  return datasetIndexTimeInMilliseconds( group, qint64( 3600 * 1000 * timeSinceGlobalReferenceInHours ) );
+}
+
+void QgsMeshDataProviderTemporalCapabilities::addGroupReferenceDateTime( int group, const QDateTime &reference )
+{
+  if ( ( !mGlobalReferenceDateTime.isValid() && reference.isValid() ) ||
+       ( reference.isValid()  && mGlobalReferenceDateTime.isValid() && reference < mGlobalReferenceDateTime ) )
+    mGlobalReferenceDateTime = reference;
+
+  mGroupsReferenceDateTime[group] = reference;
+}
+
+void QgsMeshDataProviderTemporalCapabilities::addDatasetTimeInMilliseconds( int group, qint64 time )
+{
+  QList<qint64> &datasetTimes = mDatasetTimeSinceGroupReference[group];
+  datasetTimes.append( time );
+}
+
+void QgsMeshDataProviderTemporalCapabilities::addDatasetTime( int group, double time )
+{
+  qint64 unitTimeFactor = 1;
+  switch ( mTemporalUnit )
+  {
+    case QgsUnitTypes::TemporalMilliseconds:
+      unitTimeFactor = 1;
+      break;
+    case QgsUnitTypes::TemporalSeconds:
+      unitTimeFactor = 1000;
+      break;
+    case QgsUnitTypes::TemporalMinutes:
+      unitTimeFactor = 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalHours:
+      unitTimeFactor = 60 * 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalDays:
+      unitTimeFactor = 24 * 60 * 60 * 1000;
+      break;
+    case QgsUnitTypes::TemporalWeeks:
+      unitTimeFactor = 7 * 24 * 60 * 60 * 1000;
+      break;
+    default:
+      break;
```

Comment:
@vcloarec this looks like general utlity maybe needed also by other layer types? I am wondering if it hasn't better place somewhere outside the mesh layer code?